File: frameworks/java/micronaut/UnitTests/src/test/java/unittests/ErrorsTests.java
package unittests;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.Set;

import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

/** errors: every refusal is Micronaut's own, and its default error processor writes each body. */
class ErrorsTests extends MicronautApp {

    // rb:test errors.unmatched
    @Test
    @Tag("errors.unmatched")
    void aPathNoRouteMatchesIsTheRouters404() throws Exception {
        errorBody(404, "Not Found", get("/errors/unmatched"));
    }

    // rb:test errors.not_found
    @Test
    @Tag("errors.not_found")
    void anIdWithNoRowIsTheHandlers404() throws Exception {
        errorBody(404, "Not Found", get("/items/999999"));
    }

    // rb:test errors.wrong_method
    @Test
    @Tag("errors.wrong_method")
    void aMethodThePathHasNoRouteForIs405() throws Exception {
        HttpResponse<byte[]> response = send("POST", "/items/17", "application/json", new byte[0]);

        errorBody(405, "Method Not Allowed", response);
        assertEquals(Set.of("GET", "HEAD", "PUT", "PATCH", "DELETE"), Set.of(Answer.header(response, "allow").split(",\\s*")));
    }

    // rb:test errors.malformed
    @Test
    @Tag("errors.malformed")
    void aBodyThatIsNotJsonIsRefusedWith400() throws Exception {
        byte[] body = "{\"customerId\": 1, \"lines\": [".getBytes(StandardCharsets.UTF_8);

        errorBody(400, "Bad Request", send("POST", "/body/validate/small", "application/json", body));
    }

    private static void errorBody(int status, String message, HttpResponse<byte[]> response) {
        assertEquals(status, response.statusCode());
        assertEquals("application/json", Answer.header(response, "content-type"));
        assertEquals(message, Answer.json(response).get("message").getStringValue());
    }
}

File: frameworks/dotnet/hardened/UnitTests/ErrorsTests.cs
namespace UnitTests;

public class ErrorsTests
{
    // rb:test errors.unmatched
    [ModuleTest]
    [Trait("corpus", "errors.unmatched")]
    public async Task A_path_with_no_route_is_not_found(ITestWebApp app)
    {
        TestWebResponse response = await app.Get("/errors/unmatched");

        response.Assert.NotFound();
        Assert.Empty(await response.ReadTextAsync());
    }

    // rb:test errors.wrong_method
    [ModuleTest]
    [Trait("corpus", "errors.wrong_method")]
    public async Task A_method_the_path_lacks_is_not_allowed(ITestWebApp app)
    {
        TestWebResponse response = await app.Request("POST", null, "/items/17");

        Assert.Equal(405, response.StatusCode);
        Assert.Equal("DELETE, GET, HEAD, PATCH, PUT", Answer.Header(response, "Allow"));
    }

    // rb:test errors.not_found
    [ModuleTest]
    [Trait("corpus", "errors.not_found")]
    public async Task A_row_that_is_not_there_is_not_found(ITestWebApp app)
    {
        TestWebResponse response = await app.Get("/items/999999");

        response.Assert.NotFound();
        Assert.Equal("urn:hardened:problem:not-found", (string)(await Answer.Json(response))["type"]!);
    }

    // rb:test errors.malformed
    [ModuleTest]
    [Trait("corpus", "errors.malformed")]
    public async Task A_body_that_is_not_json_is_refused_by_the_binder(ITestWebApp app)
    {
        TestWebResponse response = await app.Request("POST", null, "/body/validate/small", request => request.RawBody("""{"customerId": 1, "lines": ["""));

        response.Assert.BadRequest();
        JsonNode error = (await Answer.Json(response))["errors"]![0]!;
        Assert.Equal("order", (string)error["field"]!);
        Assert.Equal("invalid", (string)error["code"]!);
    }
}
